### PR TITLE
C++ to compute conductance

### PR DESCRIPTION
**Pull request by MengLiuPurdue**

### CODE REVIEW

File: localgraphclustering/GraphLocal.py [hunk 1]
```diff
@@ -403,30 +408,37 @@ def neighbors(self,vertex):
         """
         return self.adjacency_matrix[:,vertex].nonzero()[0].tolist()
 
-    def compute_conductance(self,R):
+    def compute_conductance(self,R,cpp=True):
```
Comment (dgleich) [inline localgraphclustering/GraphLocal.py]: Let's drop this function. The work is the same as set_scores.

File: localgraphclustering/src/lib/include/routines.hpp [hunk 2]
```diff
@@ -8,6 +8,7 @@
 #include "sparserank.hpp" // include our sorted-list functions
 #include "sparsevec.hpp" // include our sparse hashtable functions
 #include <tuple>
+#include <iostream>
```
Comment (dgleich) [inline localgraphclustering/src/lib/include/routines.hpp]: We shouldn't have iostream here.

File: localgraphclustering/src/lib/include/routines.hpp [hunk 3]
```diff
@@ -216,6 +218,26 @@ pair<itype, itype> graph<vtype,itype>::get_stats(unordered_map<vtype, vtype>& R_
     return set_stats;
 }
 
+template<typename vtype, typename itype>
+pair<double, double> graph<vtype,itype>::get_stats_weighted(unordered_map<vtype, vtype>& R_map, vtype nR)
```
Comment (dgleich) [inline localgraphclustering/src/lib/include/routines.hpp]: Any reason these are using an unordered_map vs. unordered_set ?

File: localgraphclustering/src/lib/set_scores.cpp [hunk 4]
```diff
@@ -0,0 +1,100 @@
+/* Compute cut size, number of edges given a set and a graph */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <unordered_map>
+#include <queue>
+#include <iostream>
```
Comment (dgleich) [inline localgraphclustering/src/lib/set_scores.cpp]: we should remove iostream

File: localgraphclustering/src/lib/set_scores.cpp [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+
+using namespace std;
+
+void set_scores32(
+		uint32_t n, uint32_t* ai, uint32_t* aj, uint32_t offset,
+		uint32_t* R, uint32_t nR, uint32_t* voltrue, uint32_t* cut)
+{
+	graph<uint32_t,uint32_t> g(ai[n],n,ai,aj,NULL,offset,NULL);
+    unordered_map<uint32_t, uint32_t> R_map;
+    for (size_t i = 0; i < nR; i ++) {
```
Comment (dgleich) [inline localgraphclustering/src/lib/set_scores.cpp]: This for loop is duplicated in each of the calls. Can you turn it into a function?

**Comment by dgleich** [comment]
@MengLiuPurdue the right place for the matplotlib.agg is in the test_examples() function in test_notebooks.py

File: localgraphclustering/tests/test_notebooks.py [hunk 6]
```diff
@@ -29,4 +30,76 @@ def _exec_notebook(notebook_filename):
     ep.preprocess(nb, {'metadata': {'path': 'notebooks/'}})
 
 def test_examples():
-	_exec_notebook('notebooks/examples.ipynb')
+    _exec_notebook('notebooks/examples.ipynb')
+"""
```
Comment (dgleich) [inline localgraphclustering/tests/test_notebooks.py]: We should add a link to where this code came from!

Comment (MengLiuPurdue) [reply localgraphclustering/tests/test_notebooks.py]: Ok, working on it. I also in the middle of adding more tests to get the coverage higher than 90% hopefully.

File: localgraphclustering/src/lib/set_scores.cpp [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+	pair<double, double> set_stats = g.get_stats_weighted(R_map, nR);
+	voltrue[0] = get<0>(set_stats);
+	cut[0] = get<1>(set_stats);
+}
+
+void set_scores_weighted64(
+		int64_t n, int64_t* ai, int64_t* aj, double* a, double* degrees, int64_t offset,
+		int64_t* R, int64_t nR, double* voltrue, double* cut)
+{
+	graph<int64_t,int64_t> g(ai[n],n,ai,aj,a,offset,degrees);
```
Comment (dgleich) [inline localgraphclustering/src/lib/set_scores.cpp]: Make sure you convert the tabs to spaces.

**Comment by MengLiuPurdue** [comment]
@dgleich "So it shouldn't be in the ipynb file, or in the main code... only in the test_notebooks.py file.." I tried, but that didn't seem to work.

File: localgraphclustering/cpp/__init__.py [hunk 8]
```diff
@@ -24,7 +24,13 @@ def load_library():
     lib=ctypes.cdll.LoadLibrary(find_path())
     return lib
 
+def is_loaded(lib):
```
Comment (dgleich) [inline localgraphclustering/cpp/__init__.py]: Let's remove this ,it wlll never work on windows.

File: localgraphclustering/tests/test_algs.py [hunk 9]
```diff
@@ -1,78 +1,93 @@
...
     minverts, minvals = g.local_extrema(cond,True)
     print("vertices with minimum conductance neighborhood:",minverts)
 
-def test_fiedler():
-    import numpy as np
-    g = GraphLocal("localgraphclustering/tests/data/dolphins.edges",separator=" ")
+    # Test graph with more than one components
+    G = GraphLocal("notebooks/datasets/neuro-fmri-01.edges",file_type = "edgelist", separator = " ", header = True)
+
+def setup_fiedler_test(g):
```
Comment (dgleich) [inline localgraphclustering/tests/test_algs.py]: Why change the names of these to setup?

Comment (dgleich) [reply localgraphclustering/tests/test_algs.py]: Ahh, I see now...

**Pull request merged**